### PR TITLE
Generic state implementation

### PR DESCRIPTION
**Pull request by ItzNotABug**
As discussed in here #106.

Note: I have not updated the sample apps yet. I'll update them as soon as we finalise these apis.

### CODE REVIEW

**Comment by alorma** [comment]
Please, rebase the PR do only reflects the current changes.

**Comment by ItzNotABug** [comment]
I've merged the `main` changes to the forked branch.
Should reflect the latest changes now.

**Comment by alorma** [comment]
Thanks. I will check later on (or tomorrow) , but good work there!!!

File: compose-settings-storage-preferences/src/main/kotlin/com/alorma/compose/settings/storage/preferences/PreferenceSettingValueState.kt [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+import android.content.SharedPreferences
+import androidx.compose.runtime.*
+import androidx.compose.ui.platform.LocalContext
+import androidx.core.content.edit
+import androidx.preference.PreferenceManager
+import com.alorma.compose.settings.storage.base.SettingValueState
+
+@Composable
+fun <T> rememberComposeSettingsPreference(
+  key: String,
```
[the remark this answers is not in the document]
Comment (ItzNotABug) [reply compose-settings-storage-preferences/src/main/kotlin/com/alorma/compose/settings/storage/preferences/PreferenceSettingValueState.kt:12]: I was trying to keep the library name with the function as in `remember-ComposeSettings-preference`

How about `rememberSettingsState`?

File: compose-settings-storage-preferences/src/main/kotlin/com/alorma/compose/settings/storage/preferences/PreferenceSettingValueState.kt [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+  return remember {
+      PreferenceSettingValueState(
+          key = key, defaultValue = defaultValue,
+          delimiter = delimiter, preferences = preferences
+      )
+  }
+}
+
+@Suppress("unchecked_cast")
+class PreferenceSettingValueState<T>(
```
[the remark this answers is not in the document]
Comment (ItzNotABug) [reply compose-settings-storage-preferences/src/main/kotlin/com/alorma/compose/settings/storage/preferences/PreferenceSettingValueState.kt:26]: Hmm, I guess we can move the `delimeter` to internals and keep it `private`.

File: compose-settings-storage-preferences/src/main/kotlin/com/alorma/compose/settings/storage/preferences/PreferenceSettingValueState.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (alorma) [reply compose-settings-storage-preferences/src/main/kotlin/com/alorma/compose/settings/storage/preferences/PreferenceSettingValueState.kt:12]: Much better

File: app/src/main/java/com/alorma/compose/settings/example/ui/screens/ListScreen.kt [hunk 3]
```diff
@@ -48,6 +51,9 @@ fun ListScreen(
           )
         }
       },
+      onItemSelected = { _, text ->
+        Toast.makeText(localContext, "Selected Item: $text", Toast.LENGTH_SHORT).show()
```
[the remark this answers is not in the document]
Comment (ItzNotABug) [reply app/src/main/java/com/alorma/compose/settings/example/ui/screens/ListScreen.kt]: Will refactor this impl along with the final changes if that works.

File: compose-settings-storage-preferences/src/main/kotlin/com/alorma/compose/settings/storage/preferences/PreferenceSettingValueState.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (ItzNotABug) [reply compose-settings-storage-preferences/src/main/kotlin/com/alorma/compose/settings/storage/preferences/PreferenceSettingValueState.kt:12]: Any ideas for `DataStore`?
Does any of these work?
1. `rememberDataStoreSettingsState`
2. `rememberSettingsDataStoreState`.

File: compose-settings-storage-preferences/src/main/kotlin/com/alorma/compose/settings/storage/preferences/PreferenceSettingValueState.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (alorma) [reply compose-settings-storage-preferences/src/main/kotlin/com/alorma/compose/settings/storage/preferences/PreferenceSettingValueState.kt:26]: Uhm. No. As the idea is that preference would be used outside, on another part of the app, lib must provide a way to set user prefs.


My idea goes onto split the rememberSettingsState and rememberSettingsSetState, so we can.hsbe different types.

Also for set one, maybe s function to map items to a single string (like joinToStringDoes) with a default impl.

[the remark this answers is not in the document]
Comment (ItzNotABug) [reply compose-settings-storage-preferences/src/main/kotlin/com/alorma/compose/settings/storage/preferences/PreferenceSettingValueState.kt:26]: If we create another method for handling `Set<*>`, doesn't it beat the purpose for creating a generic handler?

I mean the original api also included a separate `Set`, though it was limited to `Int` only.

Adding support for more primitives via `Set<*>` is understandable but the library currently uses a `Set` of `Int` only.

File: compose-settings-storage-preferences/src/main/kotlin/com/alorma/compose/settings/storage/preferences/PreferenceSettingValueState.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (alorma) [reply compose-settings-storage-preferences/src/main/kotlin/com/alorma/compose/settings/storage/preferences/PreferenceSettingValueState.kt:12]: - rememberSettingsState
- rememberPreferenceSettingsState
- rememberProtoXxxxSettingsState
- ...

[the remark this answers is not in the document]
Comment (ItzNotABug) [reply compose-settings-storage-preferences/src/main/kotlin/com/alorma/compose/settings/storage/preferences/PreferenceSettingValueState.kt:12]: I got these two for finals haha:
`rememberSettingsPreferenceState` &
`rememberSettingsDataStoreState`.

Similarly, we can use the proto version like - `rememberSettingsDataStoreProtoState`.

Works?

[the remark this answers is not in the document]
Comment (alorma) [reply compose-settings-storage-preferences/src/main/kotlin/com/alorma/compose/settings/storage/preferences/PreferenceSettingValueState.kt:12]: Sure.

I will continue checking tomorrow!

**Comment by alorma** [comment]
Closing because of code conflicts.

Please, reopen if you find it need.

**Pull request closed without merging**